core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlock.java

```diff
@@ -159,26 +159,38 @@ public Block copyPositions(int[] positions, int offset, int length)
     {
         checkArrayRange(positions, offset, length);
 
+        int[] newOffsets = new int[length + 1];
         int finalLength = 0;
-        for (int i = offset; i < offset + length; i++) {
-            finalLength += getSliceLength(positions[i]);
+        for (int i = 0; i < length; i++) {
+            int position = positions[offset + i];
+            finalLength += getSliceLength(position);
+            newOffsets[i + 1] = finalLength;
         }
+
         SliceOutput newSlice = Slices.allocate(finalLength).getOutput();
-        int[] newOffsets = new int[length + 1];
         boolean[] newValueIsNull = null;
-        if (valueIsNull != null) {
-            newValueIsNull = new boolean[length];
-        }
-
-        for (int i = 0; i < length; i++) {
-            int position = positions[offset + i];
-            if (!isEntryNull(position)) {
-                newSlice.writeBytes(slice, getPositionOffset(position), getSliceLength(position));
+        if (length > 0) {
```

Comment:
reverse `if`
```
if (length == 0)
```
that could probably be a first check in a method so you don't need to compute offsets